### ISSUE TITLE
Inaccurate injected JsonLocation with non-default constructor

### ISSUE THREAD

**Issue by liblit** [FasterXML/jackson-databind#3869]
**Describe the bug**

During JSON deserialization, I need to retain source coordinates for objects created by the `ObjectMapper`. I am trying to use a custom `InjectableValues` implementation to inject the parser's `JsonLocation` into `@JacksonInject`-annotated fields or constructor parameters. This works well when the constructed objects are POJOs with only the default nullary constructor. However, if I use a non-default constructor, then the `JsonLocation` seen by the `InjectableValues.findInjectableValue` method is no longer accurate.

**Version information**

jackson-databind 2.14.2

**To Reproduce**

[This repository](https://github.com/liblit/jackson-source-location-injection) demonstrates accurate locations with a POJO class, but inaccurate locations when a non-nullary constructor is added. In particular:

- [`src/main/java/SourceLocationInjector.java`](https://github.com/liblit/jackson-source-location-injection/blob/mainline/src/main/java/SourceLocationInjector.java) implements the custom `InjectableValues` logic. I was not sure which of several different `JsonLocation` values would be most suitable, so I'm logging three different candidate locations before simply returning `null`. Ideally, one of these three values would always be accurate, and I'd return that value. `Parser.getTokenLocation` seems perfect for POJO, but not for explicitly-constructed objects.

- [`src/main/java/POJO.java`](https://github.com/liblit/jackson-source-location-injection/blob/mainline/src/main/java/POJO.java) is a POJO class that stores just a deserialized `int` and an injected `JsonLocation`.

- [`src/main/java/Constructed.java`](https://github.com/liblit/jackson-source-location-injection/blob/mainline/src/main/java/Constructed.java) also stores a deserialized `int` and an injected `JsonLocation`, but adds an explicit, non-nullary constructor.

- [`src/main/java/Main.java`](https://github.com/liblit/jackson-source-location-injection/blob/mainline/src/main/java/Main.java) configures an `ObjectMapper` to use `SourceLocationInjector`. It then deserializes the same test JSON file twice: once as `POJO`, once as `Constructed`.

**Expected behavior**

For either `POJO` or `Constructed`, debugging output from `SourceLocationInjector.findInjectableValue` should report:

```text
[Source: (File); line: 2, column: 3]: token location
[Source: (File); line: 2, column: 15]: current location
[Source: (File); line: 1, column: 1]: context location
```

Unfortunately, we see the above output only for `POJO`. For `Constructed`, we instead see:

```text
[Source: (File); line: 3, column: 1]: token location
[Source: (File); line: 3, column: 2]: current location
[Source: (File); line: 1, column: 0]: context location
```

**Comment by cowtowncoder** [comment]
Yes, this is likely due to buffering that is needed when using arguments-taking constructors: in such cases `Location` cannot be retained; location given is likely location within `JsonParser` after reading and binding all values needed.

In theory it'd be possible to access and retain `Location`s of all tokens read for `TokenBuffer` but there are no current plans to do that. Although if someone wanted to work on opt-in `DeserializationFeature` to allow that, it'd be great.

**Comment by liblit** [comment]
Thanks for taking a look, @cowtowncoder. Do you have any alternate suggestions for (1) retaining location details (2) while using non-nullary constructors and (3) without invasive changes to Jackson itself? Or is that likely to be an impossible set of requirements, so I should decide which I'm willing to give up on now?

**Comment by cowtowncoder** [comment]
@liblit Unfortunately right now this might not be possible to achieve. It is not an unreasonable thing to want, but unfortunately I cannot think of a way to do it.
